src/Trakx.CoinGecko.ApiClient/CachedHttpClientHandler.cs

```diff
@@ -0,0 +1,127 @@
+        /// </summary>
+        /// <param name="request"></param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        protected override async Task<HttpResponseMessage> SendAsync(HttpRequestMessage request,
+            CancellationToken cancellationToken)
+        {
+
+            Guard.Against.Null(request, nameof(request));
+
+            await _semaphore.WaitAsync(cancellationToken).ConfigureAwait(false);
+            try
+            {
+                return TryGetOrSetRequestFromCache(request, cancellationToken);
+            }
+            finally
+            {
+                await Task.Delay(_millisecondsDelay, cancellationToken).ConfigureAwait(false);
+                _semaphore.Release(1);
+            }
+        }
+
+        /// <summary>
+        /// Any time an http client request is invoked, we need first check if the result was already cached on previous requests.
+        /// If so, we don't need to send a new request to the api.
+        /// </summary>
+        /// <param name="request">Http request that should be performed</param>
+        /// <param name="cancellationToken"></param>
+        /// <returns></returns>
+        private HttpResponseMessage TryGetOrSetRequestFromCache(HttpRequestMessage request, 
+            CancellationToken cancellationToken)
+        {
+            // Data cache is only applicable for GET operations
+            if (request.Method != HttpMethod.Get)
+            {
+                return base.SendAsync(request, cancellationToken)
+                    .ConfigureAwait(false).GetAwaiter().GetResult();
+
+            }
+
+            string content;
+            var key = $"[{request.Method}]{request.RequestUri}";
+            if (!Requests.ContainsKey(key))
+            {
+                Requests[key] = new object();
+            }
+
+            lock (Requests[key])
+            {
+                if (_cache.TryGetValue(key, out content))
+                {
+                    _logger.Debug($"Loaded {request.Method} http response to {request.RequestUri} from CACHE.");
+                    return new HttpResponseMessage(HttpStatusCode.OK)
+                    {
+                        Content = new StringContent(content)
+                    };
+                }
+
+                var response = base.SendAsync(request, cancellationToken)
+                    .ConfigureAwait(false).GetAwaiter().GetResult();
+                content = response.Content.ReadAsStringAsync(cancellationToken)
+                    .ConfigureAwait(false).GetAwaiter().GetResult();
+                var cacheExpirationOptions = new MemoryCacheEntryOptions
+                {
+                    // Set to use 'AbsoluteExpiration' as 'SlidingExpiration' was not working as expected (i.e. set to 30 seconds and it never got expired)
+                    AbsoluteExpiration = DateTime.UtcNow.AddSeconds(_apiConfiguration.CacheDurationInSeconds ?? 10),
```

Comment:
AbsoluteExpirationRelativeToNow also can be used here

Reply:
I will do.